In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

pd.set_option("display.max_columns", None)

In [3]:
prediction_df = pd.read_csv("../../05_outputs/predictions/churn_predictions.csv")

prediction_df.head()

,customer_id,churn_probability,predicted_churn,risk_level
0,50001,0.705,1,High Risk
1,50002,0.955,1,High Risk
2,50003,0.985,1,High Risk
3,50004,0.910,1,High Risk
4,50005,0.900,1,High Risk


In [8]:
import sys

sys.path.append(os.path.abspath("../preprocessing"))

from preprocess_churn import load_gold_data

gold_df = load_gold_data()

gold_df.head()

,customer_id,churn,tenure,city_tier,warehouse_to_home,hour_spend_on_app,number_of_device_registered,satisfaction_score,number_of_address,complain,order_amount_hike_from_last_year,coupon_used,order_count,day_since_last_order,cashback_amount,preferred_login_device,preferred_payment_mode,gender,preferred_order_cat,marital_status,tenure_missing_flag,warehouse_to_home_missing_flag,hour_spend_on_app_missing_flag,day_since_last_order_missing_flag,is_new_customer,low_satisfaction_flag,has_complaint,inactive_customer_flag,high_cashback_customer_flag,batch_id,loaded_at,loaded_by
0,50001,1,4.0,3,6.0,3.0,3.0,2,9.0,1,11.0,1.0,1.0,5.0,160.0,mobile phone,debit card,female,laptop & accessory,single,0,0,0,0,1,1,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
1,50002,1,0.0,1,8.0,3.0,4.0,3,7.0,1,15.0,0.0,1.0,0.0,121.0,phone,upi,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
2,50003,1,0.0,1,30.0,2.0,4.0,3,6.0,1,14.0,0.0,1.0,3.0,120.0,phone,debit card,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
3,50004,1,0.0,3,15.0,2.0,4.0,5,8.0,0,23.0,0.0,1.0,3.0,134.0,phone,debit card,male,laptop & accessory,single,0,0,0,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
4,50005,1,0.0,1,12.0,0.0,3.0,5,3.0,0,11.0,1.0,1.0,3.0,130.0,phone,cc,male,mobile,single,0,0,1,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user


In [10]:
segmentation_df = prediction_df.merge(
    gold_df,
    on="customer_id",
    how="left"
)

segmentation_df.head()

,customer_id,churn_probability,predicted_churn,risk_level,churn,tenure,city_tier,warehouse_to_home,hour_spend_on_app,number_of_device_registered,satisfaction_score,number_of_address,complain,order_amount_hike_from_last_year,coupon_used,order_count,day_since_last_order,cashback_amount,preferred_login_device,preferred_payment_mode,gender,preferred_order_cat,marital_status,tenure_missing_flag,warehouse_to_home_missing_flag,hour_spend_on_app_missing_flag,day_since_last_order_missing_flag,is_new_customer,low_satisfaction_flag,has_complaint,inactive_customer_flag,high_cashback_customer_flag,batch_id,loaded_at,loaded_by
0,50001,0.705,1,High Risk,1,4.0,3,6.0,3.0,3.0,2,9.0,1,11.0,1.0,1.0,5.0,160.0,mobile phone,debit card,female,laptop & accessory,single,0,0,0,0,1,1,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
1,50002,0.955,1,High Risk,1,0.0,1,8.0,3.0,4.0,3,7.0,1,15.0,0.0,1.0,0.0,121.0,phone,upi,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
2,50003,0.985,1,High Risk,1,0.0,1,30.0,2.0,4.0,3,6.0,1,14.0,0.0,1.0,3.0,120.0,phone,debit card,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
3,50004,0.910,1,High Risk,1,0.0,3,15.0,2.0,4.0,5,8.0,0,23.0,0.0,1.0,3.0,134.0,phone,debit card,male,laptop & accessory,single,0,0,0,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
4,50005,0.900,1,High Risk,1,0.0,1,12.0,0.0,3.0,5,3.0,0,11.0,1.0,1.0,3.0,130.0,phone,cc,male,mobile,single,0,0,1,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user


In [11]:
segmentation_df.shape

(5630, 35)

In [12]:
segmentation_df["risk_level"].value_counts()

risk_level
Low Risk       4680
High Risk       853
Medium Risk      97
Name: count, dtype: int64

In [13]:
segmentation_df["risk_level"].value_counts(normalize=True) * 100

risk_level
Low Risk       83.126110
High Risk      15.150977
Medium Risk     1.722913
Name: proportion, dtype: float64

In [14]:
high_risk_df = segmentation_df[
    segmentation_df["risk_level"] == "High Risk"
].copy()

high_risk_df.shape

(853, 35)

In [15]:
high_risk_df["complain"].value_counts(normalize=True) * 100

complain
1    55.685815
0    44.314185
Name: proportion, dtype: float64

In [16]:
high_risk_df["tenure"].describe()

count    853.000000
mean       2.631887
std        4.828808
min        0.000000
25%        0.000000
50%        1.000000
75%        1.000000
max       21.000000
Name: tenure, dtype: float64

In [17]:
high_risk_df["preferred_order_cat"].value_counts()

preferred_order_cat
mobile phone          327
mobile                198
laptop & accessory    185
fashion               118
grocery                19
others                  6
Name: count, dtype: int64

In [24]:
segmentation_df["preferred_payment_mode"] = segmentation_df["preferred_payment_mode"].replace({
    "cod": "cash on delivery",
    "cc": "credit card"
})

segmentation_df["preferred_login_device"] = segmentation_df["preferred_login_device"].replace({
    "phone": "mobile phone"
})

In [25]:
high_risk_df = segmentation_df[
    segmentation_df["risk_level"] == "High Risk"
].copy()

In [27]:
high_risk_df["preferred_payment_mode"].value_counts()

preferred_payment_mode
debit card          325
credit card         221
e wallet            133
cash on delivery    106
upi                  68
Name: count, dtype: int64

In [26]:
high_risk_df["preferred_payment_mode"].value_counts(normalize=True) * 100

preferred_payment_mode
debit card          38.100821
credit card         25.908558
e wallet            15.592028
cash on delivery    12.426729
upi                  7.971864
Name: proportion, dtype: float64

In [28]:
risk_numeric_summary = segmentation_df.groupby("risk_level")[
    [
        "tenure",
        "cashback_amount",
        "warehouse_to_home",
        "day_since_last_order",
        "order_count",
        "coupon_used",
        "satisfaction_score"
    ]
].mean()

risk_numeric_summary

,tenure,cashback_amount,warehouse_to_home,day_since_last_order,order_count,coupon_used,satisfaction_score
risk_level,,,,,,,
High Risk,2.631887,158.248535,15.798359,2.923798,2.682298,1.667057,3.406800
Low Risk,11.070726,180.725427,14.785043,4.549145,2.893590,1.667308,3.001496
Medium Risk,6.422680,175.010309,15.020619,4.134021,3.391753,1.907216,3.226804


In [29]:
os.makedirs("../../05_outputs/segments", exist_ok=True)

high_risk_df.to_csv(
    "../../05_outputs/segments/high_risk_customers.csv",
    index=False
)

os.path.exists("../../05_outputs/segments/high_risk_customers.csv")

True

## Segmentation Insights

The prediction output was merged with the Gold customer table to analyze churn risk segments.

The model classified 853 customers as High Risk, representing around 15.15% of the total customer base. Most customers were classified as Low Risk, while only a small share fell into the Medium Risk group.

High Risk customers show several clear patterns. More than half of them have submitted complaints, suggesting that complaint handling is a key retention lever.

High Risk customers also have very low tenure. The median tenure of this group is only 1, which indicates that many churn-risk customers are still in the early stage of their customer lifecycle. This suggests that onboarding and early engagement campaigns are important.

In terms of preferred order category, High Risk customers are concentrated in mobile-related categories, especially mobile phone and mobile. This may suggest that these categories are more one-off purchase driven and may require different retention strategies compared with repeat-driven categories such as grocery.

The High Risk customer segment was saved to `05_outputs/segments/high_risk_customers.csv` for further business recommendation.

## Segmentation Summary

The prediction output was merged with the Gold customer table to analyze churn risk segments.

The model classified 853 customers as High Risk, representing around 15.15% of the total customer base. Most customers were classified as Low Risk, while only a small share fell into the Medium Risk group.

High Risk customers show several clear patterns. More than half of them have submitted complaints, suggesting that complaint handling is a key retention lever.

High Risk customers also have very low tenure. The average tenure of the High Risk group is around 2.63, compared with around 11.07 for Low Risk customers. This suggests that many churn-risk customers are still in the early stage of their customer lifecycle. Onboarding and early engagement are therefore critical.

High Risk customers are concentrated in mobile-related categories, especially mobile phone and mobile. This may suggest that these categories are more one-off purchase driven and may require different retention strategies compared with repeat-driven categories such as grocery.

High Risk customers also have lower average cashback amounts than Low Risk customers. This may suggest that incentives or personalized retention offers could help improve engagement, especially for customers with low tenure and complaint history.

Some variables should be interpreted carefully. For example, satisfaction score is higher on average among High Risk customers, which does not follow the usual expectation. Therefore, satisfaction-related variables should be further validated before making strong business conclusions.

The High Risk customer segment was saved to `05_outputs/segments/high_risk_customers.csv` for further business recommendation.